ns-labeler/src/main.rs

```diff
@@ -0,0 +1,245 @@
+        prometheus_url,
+    } = Args::parse();
+
+    log_format
+        .try_init(log_level)
+        .expect("must configure logging");
+
+    info!("patching namespace {}", namespace);
+
+    let client = Client::try_default().await?;
+    let namespaces: Api<Namespace> = Api::all(client.clone());
+    let ns = namespaces.get(&namespace).await?;
+
+    let add_annotations_root = ns.metadata.annotations.filter(|x| !x.is_empty()).is_none();
+    let add_labels_root = ns.metadata.labels.filter(|x| !x.is_empty()).is_none();
+    let cni_enabled = get_cni_enabled(client, &linkerd_namespace).await?;
+    let patch = get_patch(
+        &extension,
+        prometheus_url,
+        cni_enabled,
+        add_annotations_root,
+        add_labels_root,
+    );
+    let params: PatchParams = PatchParams::apply(FIELD_MANAGER);
+    time::timeout(
+        WRITE_TIMEOUT,
+        namespaces.patch(namespace.as_str(), &params, &Patch::Json::<()>(patch)),
+    )
+    .await?
+    .map(|_| info!("successfully patched namespace"))
+    .map_err(|e| {
+        tracing::error!("failed patching namespace: {}", e);
+        anyhow!("failed patching namespace")
+    })
+}
+
+fn get_patch(
+    extension: &str,
+    prometheus_url: Option<String>,
+    cni_enabled: bool,
+    add_annotations_root: bool,
+    add_labels_root: bool,
+) -> json_patch::Patch {
+    let mut patches = Vec::new();
+
+    if add_annotations_root {
+        patches.push(Add(AddOperation {
+            path: "/metadata/annotations".to_string(),
+            value: serde_json::json!({}),
+        }));
+    }
+
+    if add_labels_root {
+        patches.push(Add(AddOperation {
+            path: "/metadata/labels".to_string(),
+            value: serde_json::json!({}),
+        }));
+    }
+
+    prometheus_url.into_iter().for_each(|url| {
+        patches.push(Add(AddOperation {
+            path: "/metadata/annotations/viz.linkerd.io~1external-prometheus".to_string(),
+            value: serde_json::Value::String(url),
+        }));
+    });
+
+    patches.push(Add(AddOperation {
+        path: "/metadata/labels/linkerd.io~1extension".to_string(),
+        value: serde_json::Value::String(extension.to_string()),
+    }));
+
+    let level = if cni_enabled {
+        "restricted"
+    } else {
+        "privileged"
+    };
+
+    patches.push(Add(AddOperation {
+        path: "/metadata/labels/pod-security.kubernetes.io~1enforce".to_string(),
+        value: serde_json::Value::String(level.to_string()),
+    }));
+
+    debug!("patch to apply: {:?}", patches);
+
+    json_patch::Patch(patches)
+}
+
+async fn get_cni_enabled(client: Client, ns: &str) -> Result<bool> {
+    let cm_api: Api<ConfigMap> = Api::namespaced(client, ns);
+    let cm = cm_api.get(LINKERD_CONFIG_CM).await?;
+    let data = cm.data.ok_or(Error::DataNotFound)?;
+    let values = data.get("values").ok_or(Error::ValuesNotFound)?;
+    let config_values: ConfigValues = serde_yaml::from_str(&values)?;
```

Comment:
I wonder if we can avoid a wrapper type by operating on the value instead (https://docs.rs/serde_yaml/latest/serde_yaml/value/enum.Value.html#method.get).

```rust
let values: serde_yaml::Value = {
    let values = data.get("values").ok_or(Error::ValuesNotFound)?;
    serde_yaml::from_str(&values)?
}

// Will either be true or false, probably won't be too bad to handle errors
serde_yaml::from_value(values["cniEnabled]"))
```

Reply:
Good call, will do :+1: